File: 2017/day8.py
#!python

import re

pattern = re.compile(r'''^(?P<reg>[a-z]+) \s
                          (?P<instr>inc|dec) \s
                          (?P<amt>-?[0-9]+) \s
                          if \s
                          (?P<other_reg>[a-z]+) \s
                          (?P<op>[!=<>]+) \s
                          (?P<cond>-?[0-9]+)
                     ''', re.X)

registers = dict()
highest = 0
ops = {
    '==': lambda val,cond: val == cond,
    '!=': lambda val,cond: val != cond,
    '<=': lambda val,cond: val <= cond,
    '>=': lambda val,cond: val >= cond,
    '>':  lambda val,cond: val > cond,
    '<':  lambda val,cond: val < cond,
}

with open('input/day8.txt') as input_lines:
    for line in input_lines:
        m = pattern.search(line)

        reg = m.group('reg')
        instruction = m.group('instr')
        val = int(m.group('amt'))
        op = m.group('op')
        other_reg = m.group('other_reg')
        cond = int(m.group('cond'))

        if other_reg not in registers:
            registers[other_reg] = 0

        if reg not in registers:
            registers[reg] = 0


        if ops[op](registers[other_reg], cond):
            if instruction == 'inc':
                registers[reg] += val
            else:
                registers[reg] -= val

        highest = max(highest, registers[reg])

    print "maximum value is %s " % max(registers.values())
    print "highest value ever seen was %s " % highest

File: 2017/day8.pl
#!perl
use v5.24;
use warnings;
use Path::Tiny qw(path);

my $pattern = qr{^([a-z]+)   \s   # register
                  (inc|dec)  \s   # instruction
                  (-?[0-9]+) \s   # amount
                  if         \s   # if
                  ([a-z]+)   \s   # other register
                  ([!=<>]+)  \s   # op
                  (-?[0-9]+)      # condition
                }x;

my %registers;
my $max = 0;
my %ops = (
  '==' => sub { $_[0] == $_[1] },
  '!=' => sub { $_[0] != $_[1] },
  '<=' => sub { $_[0] <= $_[1] },
  '>=' => sub { $_[0] >= $_[1] },
  '<'  => sub { $_[0] <  $_[1] },
  '>'  => sub { $_[0] >  $_[1] },
);

for my $line (path('input/day8.txt')->lines) {
  my ($reg, $instr, $amt, $other_reg, $op, $cond) = $line =~ /$pattern/;

  $registers{$reg} //= 0;
  $registers{$other_reg} //= 0;

  $registers{$reg} += $instr eq 'inc' ? $amt : -1 * $amt
    if $ops{$op}->($registers{$other_reg}, $cond);

  $max = $registers{$reg} if $registers{$reg} > $max;
}

my ($biggest) = sort {; $b <=> $a } values %registers;
say "maximum is $biggest";
say "highest ever is $max";
